### ISSUE TITLE
Flaky: TestGameServerReserve

### ISSUE THREAD

**Issue by markmandel** [googleforgames/agones#1276]
Haven't had a chance to work out why this is flaky, but it is. Seen it a few times across some PRs.

```
--- FAIL: TestGameServerReserve (120.77s)
    gameserver_test.go:494: 
        	Error Trace:	gameserver_test.go:494
        	Error:      	Received unexpected error:
        	            	timed out waiting for the condition
        	            	waiting for GameServer to be Reserved default/udp-serverttdbq
        	            	agones.dev/agones/test/e2e/framework.(*Framework).WaitForGameServerState
        	            		/go/src/agones.dev/agones/test/e2e/framework/framework.go:131
        	            	agones.dev/agones/test/e2e.TestGameServerReserve
        	            		/go/src/agones.dev/agones/test/e2e/gameserver_test.go:493
        	            	testing.tRunner
        	            		/usr/local/go/src/testing/testing.go:909
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1357
        	Test:       	TestGameServerReserve
    gameserver_test.go:495: 
        	Error Trace:	gameserver_test.go:495
        	Error:      	Not equal: 
        	            	expected: "Reserved"
        	            	actual  : "Ready"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-(v1.GameServerState) (len=8) "Reserved"
        	            	+(v1.GameServerState) (len=5) "Ready"
        	            	 
        	Test:       	TestGameServerReserve
```

**Comment by markmandel** [comment]
Still happening - I wonder what is going on :thinking: 

```
time="2020-01-22 01:14:44.524" level=info msg="Scaling fleet" fleet=simple-fleet-mkb2n patch="[{ \"op\": \"replace\", \"path\": \"/spec/replicas\", \"value\": 1 }]" scale=1
--- FAIL: TestGameServerReserve (123.91s)
    gameserver_test.go:494: 
        	Error Trace:	gameserver_test.go:494
        	Error:      	Received unexpected error:
        	            	timed out waiting for the condition
        	            	waiting for GameServer to be Reserved default/udp-server4f5xf
        	            	agones.dev/agones/test/e2e/framework.(*Framework).WaitForGameServerState
        	            		/go/src/agones.dev/agones/test/e2e/framework/framework.go:131
        	            	agones.dev/agones/test/e2e.TestGameServerReserve
        	            		/go/src/agones.dev/agones/test/e2e/gameserver_test.go:493
        	            	testing.tRunner
        	            		/usr/local/go/src/testing/testing.go:909
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1357
        	Test:       	TestGameServerReserve
    gameserver_test.go:495: 
        	Error Trace:	gameserver_test.go:495
        	Error:      	Not equal: 
        	            	expected: "Reserved"
        	            	actual  : "Ready"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-(v1.GameServerState) (len=8) "Reserved"
        	            	+(v1.GameServerState) (len=5) "Ready"
        	            	 
        	Test:       	TestGameServerReserve
```

**Comment by markmandel** [comment]
```
--- FAIL: TestGameServerReserve (128.47s)
    gameserver_test.go:494: 
        	Error Trace:	gameserver_test.go:494
        	Error:      	Received unexpected error:
        	            	timed out waiting for the condition
        	            	waiting for GameServer to be Reserved default/udp-servernvks2
        	            	agones.dev/agones/test/e2e/framework.(*Framework).WaitForGameServerState
        	            		/go/src/agones.dev/agones/test/e2e/framework/framework.go:131
        	            	agones.dev/agones/test/e2e.TestGameServerReserve
        	            		/go/src/agones.dev/agones/test/e2e/gameserver_test.go:493
        	            	testing.tRunner
        	            		/usr/local/go/src/testing/testing.go:909
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1357
        	Test:       	TestGameServerReserve
        	Messages:   	GameServer Name: udp-servernvks2
    gameserver_test.go:495: 
        	Error Trace:	gameserver_test.go:495
        	Error:      	Not equal: 
        	            	expected: "Reserved"
        	            	actual  : "Ready"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-(v1.GameServerState) (len=8) "Reserved"
        	            	+(v1.GameServerState) (len=5) "Ready"
        	            	 
        	Test:       	TestGameServerReserve
        	Messages:   	GameServer Name: udp-servernvks2
```

Okay, at least we can see which gameserer it was, and go look up the logs

**Comment by markmandel** [comment]
So looking at the game server and SDK logs, everything is working as expected:

```
2020-01-29 12:56:50.413 PST - 2020/01/29 20:56:50 Received packet from 34.74.237.55:50061: RESERVE
2020-01-29 12:56:50.413 PST - "Received Reserve request, adding to queue"
2020-01-29 12:56:50.418 PST - "Updating state" (Reserved)
# (10 seconds later)
2020-01-29 12:57:00.456 PST - "Received Ready request, adding to queue"
2020-01-29 12:57:00.461 PST - "Updating state": Request Ready
```
Code: https://github.com/googleforgames/agones/blob/master/examples/simple-udp/main.go#L180-L184

My theory from the above was that the Reservation timeout was too short, and within the 10 seconds, the test doesn't have enough time to see Reservation take effect.

But the weird thing is, I keep catching this error in the gameserver controller: 
`"game server container is not currently running, try again"` which is originating from [here](https://github.com/googleforgames/agones/blob/606a17859a830ff25c317345c8c771e534a22471/pkg/gameservers/controller.go#L797) - which I can't explain just yet.

Might be worth running the e2e test locally, see if we can see something. Is the e2e simple-udp crashing somehow?

**Comment by markmandel** [comment]
Normal, working test - we can see the following events:

```
Events:
  Type    Reason          Age               From                   Message
  ----    ------          ----              ----                   -------
  Normal  PortAllocation  19s               gameserver-controller  Port allocated
  Normal  Creating        19s               gameserver-controller  Pod udp-serverp7z62 created
  Normal  Scheduled       19s               gameserver-controller  Address and port populated
  Normal  RequestReady    13s               gameserver-sidecar     SDK state change
  Normal  Reserved        12s               gameserver-sidecar     SDK state change, for 10s
  Normal  Ready           2s (x2 over 13s)  gameserver-controller  SDK.Ready() complete
  Normal  Ready           2s                gameserver-controller  Deleting Pod udp-serverp7z62
```

**Comment by drichardson** [comment]
CI just hit this in build for #1360.

**Comment by markmandel** [comment]
@drichardson if you want to give it a second pair of eyes - please do. I'm struggling to work out why this is failing. All my theories haven't panned out.

**Comment by drichardson** [comment]
@markmandel Just took a really quick look. Most obvious thing to me is that the reservation message is done over UDP and so the reservation request could be dropped, and no retry request is made.

**Comment by drichardson** [comment]
> 
> 
> @markmandel Just took a really quick look. Most obvious thing to me is that the reservation message is done over UDP and so the reservation request could be dropped, and no retry request is made.

Nevermind... if that was the case the failure would be earlier and emit: *Could not message GameServer*.

**Comment by drichardson** [comment]
I made a script that runs this continuously. It ran 990 times in a row without ever hitting the failure.

The per @markmandel's suggestion on the call today, I ran the test in parallel and almost immediately hit this issue, on both the test runners. The actual/expected was different on each test runner.

## Test Runner 1 Output

```
debian build (fix_TestGameServerReserve_1276)$ ./flaky_test_runner.sh
RUNNING TEST 1
mkdir -p ~/.kube/
mkdir -p /home/doug/code/agones/build//.gocache
mkdir -p /home/doug/code/agones/build//.config/gcloud
mkdir -p ~/.helm
make ensure-image IMAGE_TAG=agones-build:eaed174c76 BUILD_TARGET=build-build-image
make[1]: Entering directory '/home/doug/code/agones/build'
make[1]: Leaving directory '/home/doug/code/agones/build'
echo "Starting e2e integration test!"
Starting e2e integration test!
docker run --rm -v /home/doug/code/agones/build//.config/gcloud:/root/.config/gcloud -v ~/.kube/:/root/.kube -v ~/.helm:/root/.helm -v /home/doug/code/agones:/go/src/agones.dev/agones -v /home/doug/code/agones/build//.gomod:/go/pkg/mod -v /home/doug/code/agones/build//.gocache:/root/.cache/go-build -e "KUBECONFIG=/root/.kube/config" -e "GO111MODULE=on" -w /go/src/agones.dev/agones  agones-build:eaed174c76 go test -mod=vendor -race -run TestGameServerReserve agones.dev/agones/test/e2e --kubeconfig /root/.kube/config \
        --gameserver-image=gcr.io/agones-images/udp-server:0.18 \
        --pullsecret=
ok      agones.dev/agones/test/e2e      15.972s
echo "Finishing e2e integration test!"
Finishing e2e integration test!
RUNNING TEST 2
mkdir -p ~/.kube/
mkdir -p /home/doug/code/agones/build//.gocache
mkdir -p /home/doug/code/agones/build//.config/gcloud
mkdir -p ~/.helm
make ensure-image IMAGE_TAG=agones-build:eaed174c76 BUILD_TARGET=build-build-image
make[1]: Entering directory '/home/doug/code/agones/build'
make[1]: Leaving directory '/home/doug/code/agones/build'
echo "Starting e2e integration test!"
Starting e2e integration test!
docker run --rm -v /home/doug/code/agones/build//.config/gcloud:/root/.config/gcloud -v ~/.kube/:/root/.kube -v ~/.helm:/root/.helm -v /home/doug/code/agones:/go/src/agones.dev/agones -v /home/doug/code/agones/build//.gomod:/go/pkg/mod -v /home/doug/code/agones/build//.gocache:/root/.cache/go-build -e "KUBECONFIG=/root/.kube/config" -e "GO111MODULE=on" -w /go/src/agones.dev/agones  agones-build:eaed174c76 go test -mod=vendor -race -run TestGameServerReserve agones.dev/agones/test/e2e --kubeconfig /root/.kube/config \
        --gameserver-image=gcr.io/agones-images/udp-server:0.18 \
        --pullsecret=
time="2020-02-27 17:32:20.052" level=info msg="Cleaning up now."
time="2020-02-27 17:32:20.247" level=info msg="Finished cleanup."
time="2020-02-27 17:32:20.307" level=info msg="GameServer created, waiting for Ready" name=udp-servern52nr
time="2020-02-27 17:32:24.402" level=info msg="GameServer Ready" name=udp-servern52nr
time="2020-02-27 17:32:39.507" level=warning msg="error retrieving gameserver" error="gameservers.agones.dev \"udp-servern52nr\" not found"
--- FAIL: TestGameServerReserve (19.32s)
    gameserver_test.go:499:
```

## Test Runner 2 Output

```
debian build (fix_TestGameServerReserve_1276)$ ./flaky_test_runner.sh
RUNNING TEST 1
mkdir -p ~/.kube/
mkdir -p /home/doug/code/agones/build//.gocache
mkdir -p /home/doug/code/agones/build//.config/gcloud
mkdir -p ~/.helm
make ensure-image IMAGE_TAG=agones-build:eaed174c76 BUILD_TARGET=build-build-image
make[1]: Entering directory '/home/doug/code/agones/build'
make[1]: Leaving directory '/home/doug/code/agones/build'
echo "Starting e2e integration test!"
Starting e2e integration test!
docker run --rm -v /home/doug/code/agones/build//.config/gcloud:/root/.config/gcloud -v ~/.kube/:/root/.kube -v ~/.helm:/root/.helm -v /home/doug/code/agones:/go/src/agones.dev/agones -v /home/doug/code/agones/build//.gomod:/go/pkg/mod -v /home/doug/code/agones/build//.gocache:/root/.cache/go-build -e "KUBECONFIG=/root/.kube/config" -e "GO111MODULE=on" -w /go/src/agones.dev/agones  agones-build:eaed174c76 go test -mod=vendor -race -run TestGameServerReserve agones.dev/agones/test/e2e --kubeconfig /root/.kube/config \
        --gameserver-image=gcr.io/agones-images/udp-server:0.18 \
        --pullsecret=
time="2020-02-27 17:32:26.746" level=info msg="Cleaning up now."
time="2020-02-27 17:32:26.954" level=info msg="Finished cleanup."
time="2020-02-27 17:32:27.017" level=info msg="GameServer created, waiting for Ready" name=udp-server7djwd
time="2020-02-27 17:32:31.101" level=info msg="GameServer Ready" name=udp-server7djwd
--- FAIL: TestGameServerReserve (19.35s)
    gameserver_test.go:499:
                Error Trace:    gameserver_test.go:499
                Error:          Received unexpected error:
                                timed out waiting for the condition
                                waiting for GameServer to be Ready default/udp-server7djwd
                                agones.dev/agones/test/e2e/framework.(*Framework).WaitForGameServerState                                        /go/src/agones.dev/agones/test/e2e/framework/framework.go:162                                   agones.dev/agones/test/e2e.TestGameServerReserve
                                        /go/src/agones.dev/agones/test/e2e/gameserver_test.go:498                                       testing.tRunner
                                        /usr/local/go/src/testing/testing.go:909
                                runtime.goexit
                                        /usr/local/go/src/runtime/asm_amd64.s:1357
                Test:           TestGameServerReserve
    gameserver_test.go:500:
                Error Trace:    gameserver_test.go:500
                Error:          Not equal:
                                expected: "Ready"
                                actual  : "Reserved"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1,2 +1,2 @@
                                -(v1.GameServerState) (len=5) "Ready"
                                +(v1.GameServerState) (len=8) "Reserved"

                Test:           TestGameServerReserve
FAIL
time="2020-02-27 17:32:46.302" level=info msg="Cleaning up now."
time="2020-02-27 17:32:46.411" level=info msg="Finished cleanup."
FAIL    agones.dev/agones/test/e2e      19.767s
FAIL
make: *** [Makefile:245: test-e2e-integration] Error 1
Test failed on iteration 1
debian build (fix_TestGameServerReserve_1276)$
```

**Comment by drichardson** [comment]
@markmandel the output you asked for

```
root@6f8eccc9ba31:/go/src/agones.dev/agones# kubectl get events -o wide | grep udp-server7djwd
14m         Normal    PortAllocation   gameserver/udp-server7djwd                                                gameserver-controller                             Port allocated                                                                                                               14m          1       udp-server7djwd.15f752b8a36da5d5
14m         Normal    Creating         gameserver/udp-server7djwd                                                gameserver-controller                             Pod udp-server7djwd created                                                                                                  14m          1       udp-server7djwd.15f752b8a4c1c4a8
14m         Normal    Scheduled        pod/udp-server7djwd                                                       default-scheduler                                 Successfully assigned default/udp-server7djwd to gke-test-cluster-default-ae8498d1-5s00                                      14m          1       udp-server7djwd.15f752b8a5a52474
14m         Normal    Scheduled        gameserver/udp-server7djwd                                                gameserver-controller                             Address and port populated                                                                                                   14m          1       udp-server7djwd.15f752b8a6b32090
14m         Normal    Pulled           pod/udp-server7djwd          spec.containers{udp-server}                  kubelet, gke-test-cluster-default-ae8498d1-5s00   Container image "gcr.io/agones-images/udp-server:0.18" already present on machine                                            14m          1       udp-server7djwd.15f752b8d4ef9cbe
14m         Normal    Created          pod/udp-server7djwd          spec.containers{udp-server}                  kubelet, gke-test-cluster-default-ae8498d1-5s00   Created container udp-server                                                                                                 14m          1       udp-server7djwd.15f752b8d7c2e6a6
14m         Normal    Started          pod/udp-server7djwd          spec.containers{udp-server}                  kubelet, gke-test-cluster-default-ae8498d1-5s00   Started container udp-server                                                                                                 14m          1       udp-server7djwd.15f752b8e20c034b
14m         Normal    Pulling          pod/udp-server7djwd          spec.containers{agones-gameserver-sidecar}   kubelet, gke-test-cluster-default-ae8498d1-5s00   Pulling image "gcr.io/doug-agones-test/agones-sdk:1.4.0-946f84e"                                                             14m          1       udp-server7djwd.15f752b8e228f83a
14m         Normal    Pulled           pod/udp-server7djwd          spec.containers{agones-gameserver-sidecar}   kubelet, gke-test-cluster-default-ae8498d1-5s00   Successfully pulled image "gcr.io/doug-agones-test/agones-sdk:1.4.0-946f84e"                                                 14m          1       udp-server7djwd.15f752b8eebbad99
14m         Normal    Created          pod/udp-server7djwd          spec.containers{agones-gameserver-sidecar}   kubelet, gke-test-cluster-default-ae8498d1-5s00   Created container agones-gameserver-sidecar                                                                                  14m          1       udp-server7djwd.15f752b8f2430e9b
14m         Normal    Started          pod/udp-server7djwd          spec.containers{agones-gameserver-sidecar}   kubelet, gke-test-cluster-default-ae8498d1-5s00   Started container agones-gameserver-sidecar                                                                                  14m          1       udp-server7djwd.15f752b8fb87c8f9
14m         Normal    RequestReady     gameserver/udp-server7djwd                                                gameserver-sidecar                                SDK state change                                                                                                             14m          1       udp-server7djwd.15f752b9059fd7f3
14m         Normal    Ready            gameserver/udp-server7djwd                                                gameserver-controller                             SDK.Ready() complete                                                                                                         14m          1       udp-server7djwd.15f752b929f3205e
14m         Normal    Reserved         gameserver/udp-server7djwd                                                gameserver-sidecar                                SDK state change, for 10s                                                                                                    14m          1       udp-server7djwd.15f752b998c137c6
14m         Normal    Reserved         gameserver/udp-server7djwd                                                gameserver-controller                             Deleting Pod udp-server7djwd                                                                                                 14m          1       udp-server7djwd.15f752bb97715de2
14m         Normal    Killing          pod/udp-server7djwd          spec.containers{udp-server}                  kubelet, gke-test-cluster-default-ae8498d1-5s00   Stopping container udp-server                                                                                                14m          1       udp-server7djwd.15f752bb978f57ab
14m         Normal    Killing          pod/udp-server7djwd          spec.containers{agones-gameserver-sidecar}   kubelet, gke-test-cluster-default-ae8498d1-5s00   Stopping container agones-gameserver-sidecar                                                                                 14m          1       udp-server7djwd.15f752bb979400a8
```

**Comment by drichardson** [comment]
@markmandel wondered on slack if this could be a load issue. While running 3 instances of the tests, here's what my GKE web console showed for each node pool:

![system](https://user-images.githubusercontent.com/36176/75474515-f8371480-594b-11ea-8ad2-4583ffb3af75.png)
![default](https://user-images.githubusercontent.com/36176/75474517-f8cfab00-594b-11ea-9514-17c84cebccb9.png)
![metrics](https://user-images.githubusercontent.com/36176/75474518-f8cfab00-594b-11ea-94e7-5cede22e7176.png)

**Comment by drichardson** [comment]
After running the 3 instances of the load tests I used for the screenshots in my previous comment, 2 of the tests failed with a different error message (1 and 2), while test 3 failed with the same error message (about Ready != Reserved) we have already seen.

[test runner 1 output.txt](https://github.com/googleforgames/agones/files/4263387/test.runner.1.output.txt)
[test runner 2 output.txt](https://github.com/googleforgames/agones/files/4263388/test.runner.2.output.txt)
[test runner 3 output.txt](https://github.com/googleforgames/agones/files/4263389/test.runner.3.output.txt)

**Comment by drichardson** [comment]
@markmandel These look suspicious:

```
27m         Warning   Failed           pod/udp-serverhsdzc          spec.containers{udp-server}                  kubelet, gke-test-cluster-default-ae8498d1-nxlr   Error: cannot find volume "empty" to mount into container "udp-server"                                             27m          1       udp-serverhsdzc.15f75522b8fb5139
27m         Normal    Pulling          pod/udp-serverhsdzc          spec.containers{agones-gameserver-sidecar}   kubelet, gke-test-cluster-default-ae8498d1-nxlr   Pulling image "gcr.io/doug-agones-test/agones-sdk:1.4.0-946f84e"                                                   27m          1       udp-serverhsdzc.15f75522b917b232
27m         Normal    Pulled           pod/udp-serverhsdzc          spec.containers{agones-gameserver-sidecar}   kubelet, gke-test-cluster-default-ae8498d1-nxlr   Successfully pulled image "gcr.io/doug-agones-test/agones-sdk:1.4.0-946f84e"                                       27m          1       udp-serverhsdzc.15f75522c56ef769
27m         Warning   Failed           pod/udp-serverhsdzc          spec.containers{agones-gameserver-sidecar}   kubelet, gke-test-cluster-default-ae8498d1-nxlr   Error: cannot find volume "agones-sdk-token-2h5nd" to mount into container "agones-gameserver-sidecar"  
```

**Comment by drichardson** [comment]
Another e2e test in gameserver_test failed today on a documentation change.

https://console.cloud.google.com/cloud-build/builds/32b736a7-c87b-4e00-8465-d9ce19e3ea91?project=agones-images

```
--- FAIL: TestUnhealthyGameServersWithoutFreePorts (153.06s)
    gameserver_test.go:202: 
        	Error Trace:	gameserver_test.go:202
        	Error:      	Received unexpected error:
        	            	timed out waiting for the condition
        	            	waiting for GameServer to be Unhealthy default/udp-serverhswrq
        	            	agones.dev/agones/test/e2e/framework.(*Framework).WaitForGameServerState
        	            		/go/src/agones.dev/agones/test/e2e/framework/framework.go:162
        	            	agones.dev/agones/test/e2e.TestUnhealthyGameServersWithoutFreePorts
        	            		/go/src/agones.dev/agones/test/e2e/gameserver_test.go:201
        	            	testing.tRunner
        	            		/usr/local/go/src/testing/testing.go:909
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1357
        	Test:       	TestUnhealthyGameServersWithoutFreePorts
```

**Comment by markmandel** [comment]
@drichardson thanks for keeping track :+1: , but I think reporting it each time is going to be just noise at this stage. The error reports aren't showing anything new. But if we find new information, let's continue to keep track of that here for sure.

**Comment by drichardson** [comment]
It's a different test.

**Comment by markmandel** [comment]
Oh I see, sorry - totally missed that!

Let's file a new ticket for it if it continues to be an issue. Best to track these things seperately.

**Comment by drichardson** [comment]
Since we don't know what the underlying problem of this failure is, and it is an e2e test, I thought it might be a clue.

But yes, I'll file an issue for your commit that failed CI ;)

 #1376

**Comment by drichardson** [comment]
This test is so unreliable that it is useless. If there was actually a problem here, we'd ignore it as a result of a flaky test. It should be disabled until fixed.

**Comment by markmandel** [comment]
Okay, we got some debug! The TL;DR - the state never changes on Polling, but I can see an event change with state. Digging in.

```
Step #22: time="2020-03-06 20:20:04.495" level=info msg="sending RESERVE command" test=TestGameServerReserve
Step #22: time="2020-03-06 20:20:04.570" level=info msg="Received response" test=TestGameServerReserve
Step #22: time="2020-03-06 20:20:14.890" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-servermvtlv test=TestGameServerReserve
Step #22: time="2020-03-06 20:20:25.329" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-servermvtlv test=TestGameServerReserve
Step #22: time="2020-03-06 20:20:35.087" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-servermvtlv test=TestGameServerReserve
Step #22: time="2020-03-06 20:20:44.322" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-servermvtlv test=TestGameServerReserve
Step #22: time="2020-03-06 20:20:51.472" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-servermvtlv test=TestGameServerReserve
Step #22: time="2020-03-06 20:20:57.952" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-servermvtlv test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:03.907" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-servermvtlv test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:09.563" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-servermvtlv test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:14.347" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-servermvtlv test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:18.606" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-servermvtlv test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:22.844" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-servermvtlv test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:27.212" level=info msg="Waiting for states to match" awaitingState=Ready currentState=Ready gs=udp-servermvtlv test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:27.993" level=info msg="gs event details" count=1 first-time="2020-03-06 20:19:33 +0000 UTC" last-time="2020-03-06 20:19:33 +0000 UTC" message="Port allocated" name=udp-servermvtlv.15f9d07bba93024c reason=PortAllocation test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:27.993" level=info msg="gs event details" count=1 first-time="2020-03-06 20:19:34 +0000 UTC" last-time="2020-03-06 20:19:34 +0000 UTC" message="Pod udp-servermvtlv created" name=udp-servermvtlv.15f9d07bca76337a reason=Creating test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:27.993" level=info msg="gs event details" count=1 first-time="2020-03-06 20:19:34 +0000 UTC" last-time="2020-03-06 20:19:34 +0000 UTC" message="Address and port populated" name=udp-servermvtlv.15f9d07bd14a98d2 reason=Scheduled test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:27.993" level=info msg="gs event details" count=1 first-time="2020-03-06 20:19:56 +0000 UTC" last-time="2020-03-06 20:19:56 +0000 UTC" message="SDK state change" name=udp-servermvtlv.15f9d08116a79858 reason=RequestReady test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:27.994" level=info msg="gs event details" count=2 first-time="2020-03-06 20:20:03 +0000 UTC" last-time="2020-03-06 20:20:14 +0000 UTC" message="SDK.Ready() complete" name=udp-servermvtlv.15f9d082a643598f reason=Ready test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:27.994" level=info msg="gs event details" count=1 first-time="2020-03-06 20:20:04 +0000 UTC" last-time="2020-03-06 20:20:04 +0000 UTC" message="SDK state change, for 10s" name=udp-servermvtlv.15f9d082e256f080 reason=Reserved test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:28.147" level=info msg="Pod Status" status="{Running [{Initialized True 0001-01-01 00:00:00 +0000 UTC 2020-03-06 20:19:34 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2020-03-06 20:19:52 +0000 UTC  } {ContainersReady True 0001-01-01 00:00:00 +0000 UTC 2020-03-06 20:19:52 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2020-03-06 20:19:34 +0000 UTC  }]    10.138.15.244 10.40.1.134 2020-03-06 20:19:34 +0000 UTC [] [{agones-gameserver-sidecar {nil &ContainerStateRunning{StartedAt:2020-03-06 20:19:51 +0000 UTC,} nil} {nil nil nil} true 0 gcr.io/agones-images/agones-sdk:1.5.0-0691bdb docker-pullable://gcr.io/agones-images/agones-sdk@sha256:82f41e0b6c0a38e124a5b2c08756304501281db54978fc080b96cb88617a46a0 docker://d8cca6932b99ad097ebc2622edd3eb13961fe9889ccc4dbe63a8f24f6119d349} {udp-server {nil &ContainerStateRunning{StartedAt:2020-03-06 20:19:39 +0000 UTC,} nil} {nil nil nil} true 0 gcr.io/agones-images/udp-server:0.18 docker-pullable://gcr.io/agones-images/udp-server@sha256:fa64d524c817cbb8f23c59214c51b039148358bd50623ec14843e5fc83284a25 docker://d9f02f7346ad8d893b636509d53c51b2316b9c64a008eb4b96070dc8af012c78}] Burstable}" test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:29.259" level=info msg="pod event details" count=1 first-time="2020-03-06 20:19:34 +0000 UTC" last-time="2020-03-06 20:19:34 +0000 UTC" message="Successfully assigned default/udp-servermvtlv to gke-e2e-test-cluster-default-8400923d-3nsj" name=udp-servermvtlv.15f9d07bcc13ec2a reason=Scheduled test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:29.259" level=info msg="pod event details" count=1 first-time="2020-03-06 20:19:36 +0000 UTC" last-time="2020-03-06 20:19:36 +0000 UTC" message="Container image \"gcr.io/agones-images/udp-server:0.18\" already present on machine" name=udp-servermvtlv.15f9d07c4360c089 reason=Pulled test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:29.260" level=info msg="pod event details" count=1 first-time="2020-03-06 20:19:37 +0000 UTC" last-time="2020-03-06 20:19:37 +0000 UTC" message="Created container udp-server" name=udp-servermvtlv.15f9d07cb193fa05 reason=Created test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:29.260" level=info msg="pod event details" count=1 first-time="2020-03-06 20:19:40 +0000 UTC" last-time="2020-03-06 20:19:40 +0000 UTC" message="Started container udp-server" name=udp-servermvtlv.15f9d07d3d5e0934 reason=Started test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:29.260" level=info msg="pod event details" count=1 first-time="2020-03-06 20:19:40 +0000 UTC" last-time="2020-03-06 20:19:40 +0000 UTC" message="Pulling image \"gcr.io/agones-images/agones-sdk:1.5.0-0691bdb\"" name=udp-servermvtlv.15f9d07d3d92e3e5 reason=Pulling test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:29.260" level=info msg="pod event details" count=1 first-time="2020-03-06 20:19:44 +0000 UTC" last-time="2020-03-06 20:19:44 +0000 UTC" message="Successfully pulled image \"gcr.io/agones-images/agones-sdk:1.5.0-0691bdb\"" name=udp-servermvtlv.15f9d07e3496638c reason=Pulled test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:29.260" level=info msg="pod event details" count=1 first-time="2020-03-06 20:19:46 +0000 UTC" last-time="2020-03-06 20:19:46 +0000 UTC" message="Created container agones-gameserver-sidecar" name=udp-servermvtlv.15f9d07ebca549ae reason=Created test=TestGameServerReserve
Step #22: time="2020-03-06 20:21:29.261" level=info msg="pod event details" count=1 first-time="2020-03-06 20:19:51 +0000 UTC" last-time="2020-03-06 20:19:51 +0000 UTC" message="Started container agones-gameserver-sidecar" name=udp-servermvtlv.15f9d07fd2e8f8f8 reason=Started test=TestGameServerReserve

--- FAIL: TestGameServerReserve (119.40s)
    gameserver_test.go:499:
        	Error Trace:	gameserver_test.go:499
        	Error:      	Received unexpected error:
        	            	timed out waiting for the condition
        	            	waiting for GameServer to be Reserved default/udp-servermvtlv
        	            	agones.dev/agones/test/e2e/framework.(*Framework).WaitForGameServerStateWithLogger
        	            		/go/src/agones.dev/agones/test/e2e/framework/framework.go:166
        	            	agones.dev/agones/test/e2e.TestGameServerReserve
        	            		/go/src/agones.dev/agones/test/e2e/gameserver_test.go:498
        	            	testing.tRunner
        	            		/usr/local/go/src/testing/testing.go:909
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1357
        	Test:       	TestGameServerReserve
        	Messages:   	GameServer Name: udp-servermvtlv
    gameserver_test.go:500:
        	Error Trace:	gameserver_test.go:500
        	Error:      	Not equal:
        	            	expected: "Reserved"
        	            	actual  : "Ready"

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-(v1.GameServerState) (len=8) "Reserved"
        	            	+(v1.GameServerState) (len=5) "Ready"

        	Test:       	TestGameServerReserve
        	Messages:   	GameServer Name: udp-servermvtlv
```

**Comment by markmandel** [comment]
I just saw something I hadn't noticed before in the logs

```
time="2020-03-09 16:48:51.881" level=info msg="Received response" test=TestGameServerReserve
time="2020-03-09 16:49:02.257" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-serverv5tcm test=TestGameServerReserve
```
Look at the timestamps - there is a > 10 second gap there. 

Going by the event logs, the move to Reserved happens at `2020-03-09 16:48:51`, and then the return to Ready happens at `2020-03-09 16:48:49`, which is BEFORE the test to check if the GameServer is Reserved.

I don't know *why* there is a 10 second gap - but that does explain what is going on.

**Comment by roberthbailey** [comment]
`2020-03-09 16:48:49` is _before_ `2020-03-09 16:48:51`, which implies that the "return" to Ready isn't a return; it looks more like the server being ready and then transitioning to Reserved three seconds later (and the transition time lines up with the log time for the event receipt). Event logs can sometimes be coalesced; is there a way for you to tell if there was a second transition from Reserved -> Ready?

**Comment by markmandel** [comment]
I track both in the logs - you can see the count=2, and different first-time and last-time timestamps (you can get a breakdown of individual event series item, but since there is just two, then I just grabbed those two values since they were easy to get to).

**Comment by markmandel** [comment]
Here's another breakdown, with accompanying logs (slightly cleaned up to make things easier to see)

```
time="2020-03-13 22:57:25.286" level=info msg="sending RESERVE command"
time="2020-03-13 22:57:25.291" level=info msg="Received response"
time="2020-03-13 22:57:36.046" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-serverk8f4m
time="2020-03-13 22:57:46.292" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-serverk8f4m
time="2020-03-13 22:57:56.087" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-serverk8f4m
time="2020-03-13 22:58:05.092" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-serverk8f4m
time="2020-03-13 22:58:12.392" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-serverk8f4m
time="2020-03-13 22:58:19.260" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-serverk8f4m
time="2020-03-13 22:58:25.199" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-serverk8f4m
time="2020-03-13 22:58:31.158" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-serverk8f4m
time="2020-03-13 22:58:36.609" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-serverk8f4m
time="2020-03-13 22:58:40.770" level=info msg="Waiting for states to match" awaitingState=Reserved currentState=Ready gs=udp-serverk8f4m
time="2020-03-13 22:58:44.542" level=info msg="Waiting for states to match" awaitingState=Ready currentState=Ready gs=udp-serverk8f4m
time="2020-03-13 22:58:45.879" level=info msg="gs event details" count=1 first-time="2020-03-13 22:56:54 +0000 UTC" last-time="2020-03-13 22:56:54 +0000 UTC" message="Port allocated" name=udp-serverk8f4m.15fbff21c8f101ae reason=PortAllocation
time="2020-03-13 22:58:45.879" level=info msg="gs event details" count=1 first-time="2020-03-13 22:56:54 +0000 UTC" last-time="2020-03-13 22:56:54 +0000 UTC" message="Pod udp-serverk8f4m created" name=udp-serverk8f4m.15fbff21d1fbbfe0 reason=Creating
time="2020-03-13 22:58:45.879" level=info msg="gs event details" count=1 first-time="2020-03-13 22:56:54 +0000 UTC" last-time="2020-03-13 22:56:54 +0000 UTC" message="Address and port populated" name=udp-serverk8f4m.15fbff21d5fe36f1 reason=Scheduled
time="2020-03-13 22:58:45.879" level=info msg="gs event details" count=2 first-time="2020-03-13 22:57:19 +0000 UTC" last-time="2020-03-13 22:57:35 +0000 UTC" message="SDK state change" name=udp-serverk8f4m.15fbff2799164f39 reason=RequestReady
time="2020-03-13 22:58:45.879" level=info msg="gs event details" count=2 first-time="2020-03-13 22:57:24 +0000 UTC" last-time="2020-03-13 22:57:36 +0000 UTC" message="SDK.Ready() complete" name=udp-serverk8f4m.15fbff28dd1582e9 reason=Ready
time="2020-03-13 22:58:45.880" level=info msg="gs event details" count=1 first-time="2020-03-13 22:57:25 +0000 UTC" last-time="2020-03-13 22:57:25 +0000 UTC" message="SDK state change, for 10s" name=udp-serverk8f4m.15fbff28f42c59fb reason=Reserved
time="2020-03-13 22:58:45.922" level=info msg="Pod Status" status="{Running [{Initialized True 0001-01-01 00:00:00 +0000 UTC 2020-03-13 22:56:54 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2020-03-13 22:57:21 +0000 UTC  } {ContainersReady True 0001-01-01 00:00:00 +0000 UTC 2020-03-13 22:57:21 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2020-03-13 22:56:54 +0000 UTC  }]    10.138.15.239 10.40.4.124 2020-03-13 22:56:54 +0000 UTC [] [{agones-gameserver-sidecar {nil &ContainerStateRunning{StartedAt:2020-03-13 22:57:15 +0000 UTC,} nil} {nil nil nil} true 0 gcr.io/agones-images/agones-sdk:1.5.0-98037d2 docker-pullable://gcr.io/agones-images/agones-sdk@sha256:56c74f2d586b02d7b15fec664581c6884b4e565df87fb265f81f81845c2a0656 docker://cc086671bfade8901c4c6835ab488dfdd644df2dea5e1fc80af7a5192e4c0c97} {udp-server {nil &ContainerStateRunning{StartedAt:2020-03-13 22:57:09 +0000 UTC,} nil} {nil nil nil} true 0 gcr.io/agones-images/udp-server:0.18 docker-pullable://gcr.io/agones-images/udp-server@sha256:fa64d524c817cbb8f23c59214c51b039148358bd50623ec14843e5fc83284a25 docker://4d1224024b4c7d572ea21e43a3a5522aa7a8d868f66949d2d21833dc23a53de6}] Burstable}"
time="2020-03-13 22:58:47.919" level=info msg="gs pod details" count=1 first-time="2020-03-13 22:56:54 +0000 UTC" last-time="2020-03-13 22:56:54 +0000 UTC" message="Successfully assigned default/udp-serverk8f4m to gke-e2e-test-cluster-default-8400923d-xxlb" name=udp-serverk8f4m.15fbff21d38e6997 reason=Scheduled
time="2020-03-13 22:58:47.919" level=info msg="gs pod details" count=1 first-time="2020-03-13 22:57:04 +0000 UTC" last-time="2020-03-13 22:57:04 +0000 UTC" message="Container image \"gcr.io/agones-images/udp-server:0.18\" already present on machine" name=udp-serverk8f4m.15fbff24094b5015 reason=Pulled
time="2020-03-13 22:58:47.919" level=info msg="gs pod details" count=1 first-time="2020-03-13 22:57:05 +0000 UTC" last-time="2020-03-13 22:57:05 +0000 UTC" message="Created container udp-server" name=udp-serverk8f4m.15fbff246085b8d7 reason=Created
time="2020-03-13 22:58:47.920" level=info msg="gs pod details" count=1 first-time="2020-03-13 22:57:09 +0000 UTC" last-time="2020-03-13 22:57:09 +0000 UTC" message="Started container udp-server" name=udp-serverk8f4m.15fbff2533e04931 reason=Started
time="2020-03-13 22:58:47.920" level=info msg="gs pod details" count=1 first-time="2020-03-13 22:57:09 +0000 UTC" last-time="2020-03-13 22:57:09 +0000 UTC" message="Pulling image \"gcr.io/agones-images/agones-sdk:1.5.0-98037d2\"" name=udp-serverk8f4m.15fbff2533fb18a5 reason=Pulling
time="2020-03-13 22:58:47.920" level=info msg="gs pod details" count=1 first-time="2020-03-13 22:57:09 +0000 UTC" last-time="2020-03-13 22:57:09 +0000 UTC" message="Successfully pulled image \"gcr.io/agones-images/agones-sdk:1.5.0-98037d2\"" name=udp-serverk8f4m.15fbff253f74943c reason=Pulled
time="2020-03-13 22:58:47.920" level=info msg="gs pod details" count=1 first-time="2020-03-13 22:57:10 +0000 UTC" last-time="2020-03-13 22:57:10 +0000 UTC" message="Created container agones-gameserver-sidecar" name=udp-serverk8f4m.15fbff2564b08f79 reason=Created
time="2020-03-13 22:58:47.920" level=info msg="gs pod details" count=1 first-time="2020-03-13 22:57:15 +0000 UTC" last-time="2020-03-13 22:57:15 +0000 UTC" message="Started container agones-gameserver-sidecar" name=udp-serverk8f4m.15fbff26b5a7ce17 reason=Started
```

- 22:57:24     - First move to Ready state
- 22:57:25.291 - Receive Response on RESERVE command
- 22:57:25     - Move to Reserved state 
- 22:57:36     - Second move to Ready state
- 22:57:36.046 - start testing for states to match. (over 10s later)

It seems odd that the start of the test kicks in almost always just over 10s later.

**Comment by markmandel** [comment]
Thought I had a solution but https://github.com/googleforgames/agones/pull/1412#issuecomment-599137759 proves otherwise.

> The salient line, check out this gap:
```
time="2020-03-14 21:21:08.945" level=info msg="checking for reserved state"
time="2020-03-14 21:21:19.310" level=info msg="Waiting for reserved state" currentState=Ready gs=udp-serverkhnmx
```

> Basically, it seems like `gs, err = framework.AgonesClient.AgonesV1().GameServers(gs.Namespace).Get(gs.Name, metav1.GetOptions{}) ` is getting blocked, for ~ 10 seconds (which is weird it's around that time), (maybe the API server is getting overwhelmed???) and it's actually nothing to do with the goroutine.

Next stage - I'm going to wait 10 seconds, then check the Event stream that it is as expected. I can't come up with a better solution that than.

**Comment by dzmitry-lahoda** [comment]
Hi, not found a better place.

Agones 1.21-rc. K8s 1.21. Windows Core GS and Windows Nano GS(simple GS). Looped may be 5-6 times like this
<img width="1103" alt="image" src="https://user-images.githubusercontent.com/757125/154946205-23aaed3a-95f5-4dfa-bbcf-28ab659d9f9a.png">

After that successfully started. Maybe it was our cluster issue. Not agones.
